### PR TITLE
New Xenomorph Caste: Trapper

### PR DESCRIPTION
**Pull request by jjpark-kb**
## About The Pull Request
adds a new caste: trapper, t2, from drone
trapper has three "information" abilities: sensor, tripwire, and zoom
sensor places a sensor that notifies the hive if there is a host nearby
tripwire places a tripwire that notifies the hive if a host trips the tripwire
zoom is just the queen zoom

makes carrier t3

## Why It's Good For The Game
more information based and more trap based xenos rather than the hit and run tactic that is commonly performed. Can create interesting scenarios where xenos could glean information and try to coordinate ambushes and traps on unsuspecting marines.

## Changelog
:cl:
add: Added new things
add: Added more things
del: Removed old things
tweak: tweaked a few things
balance: rebalanced something
fix: fixed a few things
soundadd: added a new sound thingy
sounddel: removed an old sound thingy
imageadd: added some icons and images
imagedel: deleted some icons and images
spellcheck: fixed a few typos
code: changed some code
refactor: refactored some code
config: changed some config setting
admin: messed with admin stuff
server: something server ops should know
/:cl:

### CODE REVIEW

**Comment by jroinc** [comment]
What does this add that the hivemind doesn't already have?

**Comment by jjpark-kb** [comment]
> What does this add that the hivemind doesn't already have?

this is a draft PR, so things are subject to change.
Hivemind cant directly fight. Sure the hivemind can give out information, but the hivemind is only one xeno, and so isnt omnipotent. These "sensor traps" give information to the whole hive automatically without needing someone at the location.
 Again, draft pr, so the trapper will be developed some more.

File: code/game/objects/structures/xeno.dm [hunk 1]
```diff
@@ -718,3 +718,76 @@ TUNNEL
...
+	var/area/A = get_area(src)
+	for(var/X in GLOB.alive_xeno_list)
+		to_chat(X, "<span class='xenowarning'>A sensor has been destroyed at [A].</span>")
+	. = ..()
+
+/obj/effect/alien/resin/sensor/process()
+	for(var/mob/living/carbon/H in GLOB.alive_human_list)
+		if(cooldown)
+			return
+		if(H in viewers(src))
```
Comment (psykzz) [inline code/game/objects/structures/xeno.dm:750]: You need to rework this so you don't call viewers for every alive human

File: code/game/objects/structures/xeno.dm [hunk 2]
```diff
@@ -718,3 +718,76 @@ TUNNEL
...
+			if(!isxeno(H) && H.stat != DEAD)
+				to_chat(H, "<span class='warning'>You feel something glance at you...</span>")
+				number++
+	var/area/A = get_area(src)
+	for(var/X in GLOB.alive_xeno_list)
+		if(number > 1)
+			var/textnumber = num2text(number)
+			to_chat(X, "<span class='xenowarning'>There are [textnumber] hosts at [A].</span>")
+		else
+			to_chat(X, "<span class='xenowarning'>There is a host at [A].</span>")
```
Comment (psykzz) [inline code/game/objects/structures/xeno.dm:760]: This doest't check for nearby at all

File: code/game/objects/structures/xeno.dm [hunk 3]
```diff
@@ -718,3 +718,76 @@ TUNNEL
...
+	opacity = FALSE
+	anchored = TRUE
+	max_integrity = 5
+	layer = RESIN_STRUCTURE_LAYER
+
+/obj/effect/alien/resin/tripwire/Initialize()
+	. = ..()
+
+/obj/effect/alien/resin/tripwire/Destroy()
+	. = ..()
```
Comment (psykzz) [inline code/game/objects/structures/xeno.dm:785]: No need to add these if they dont do anything

File: code/game/objects/structures/xeno.dm (hunk 2 above)
Comment (jjpark-kb) [reply code/game/objects/structures/xeno.dm:760]: it doesnt check for nearby intentionally as its meant to broadcast to the whole hive. Should I make it only in range?

File: code/game/objects/structures/xeno.dm (hunk 3 above)
Comment (jjpark-kb) [reply code/game/objects/structures/xeno.dm:785]: alrighty, Ill remove those

File: code/game/objects/structures/xeno.dm (hunk 1 above)
Comment (jjpark-kb) [reply code/game/objects/structures/xeno.dm:750]: What should I do? Should I do the for on the viewers instead and then check if they are alive?

Comment (jjpark-kb) [reply code/game/objects/structures/xeno.dm]: done

File: code/game/objects/structures/xeno.dm (hunk 3 above)
Comment (jjpark-kb) [reply code/game/objects/structures/xeno.dm]: done!

File: code/modules/mob/living/carbon/xenomorph/castes/trapper/trapper.dm [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+	upgrade = XENO_UPGRADE_ZERO
+	pixel_x = -16  //Needed for 2x2
+	old_x = -16
+	inherent_verbs = list(
+		/mob/living/carbon/xenomorph/proc/vent_crawl,
+		)
+	var/list/tunnels = list() //list of active tunnels
+	var/mob/living/carbon/xenomorph/observed_xeno //the Xenomorph the trapper is currently overwatching (no overwatch, but required for zoom)
+
+/mob/living/carbon/xenomorph/panther/set_stat()
```
Comment (spookydonut) [inline code/modules/mob/living/carbon/xenomorph/castes/trapper/trapper.dm:21]: panther?

Comment (jjpark-kb) [reply code/modules/mob/living/carbon/xenomorph/castes/trapper/trapper.dm:21]: Let me correct that. Thank you!

File: code/game/objects/structures/xeno.dm (hunk 1 above)
Comment (jjpark-kb) [reply code/game/objects/structures/xeno.dm]: Solved

**Comment by terranaut1** [comment]
Some maps have pretty big and poorly defined areas, how do you plan to give a player precise information that he can act on or pass on in good faith without him being there himself like a Hivemind would?

**Comment by jroinc** [comment]
> Some maps have pretty big and poorly defined areas, how do you plan to give a player precise information that he can act on or pass on in good faith without him being there himself like a Hivemind would?

Only one that's really that bad is prison, most of the others are decent.

"X detected at South Central Barrens" is actionable.
"X detected in Lower Med Security Cells" is not.

**Comment by terranaut1** [comment]
the entire caves of icy (like 60% of the map or so) is defined via 3 areas

File: code/game/objects/structures/xeno.dm (hunk 2 above)
Comment (spookydonut) [reply code/game/objects/structures/xeno.dm:757]: don't use process, this won't work like you want.

**Comment by jjpark-kb** [comment]
Instead of process, how should I proceed? It currently checks every second with the process and then if it finds a host, will run the stuff and then stop processing for 40 seconds. After the callback successfully works, it will then turn back on the process and check every 1 second again. This is what I was looking for. Is there a better way of doing it?

**Comment by jjpark-kb** [comment]
after consideration, "trap gameplay bad".

**Pull request closed without merging**